plugin-java/src/main/java/appland/execution/AppMapJavaPackageConfig.java

```diff
@@ -220,8 +220,13 @@ private static VirtualFile[] collectSourceRootsWithDependencies(@NotNull Module
                 .flatMap(m -> Arrays.stream(ModuleRootManager.getInstance(m).getContentRoots()))
                 .collect(Collectors.toUnmodifiableSet());
 
-        return Arrays.stream(ProjectRootManager.getInstance(module.getProject()).getContentSourceRoots())
-                .filter(sourceRoot -> VfsUtilCore.isUnder(sourceRoot, contentRoots))
+        ProjectRootManager rootManager = ProjectRootManager.getInstance(module.getProject());
```

Comment:
The new code is filtering all source roots of the project, which means that all source roots of the project are returned. With all source roots, all packages available in the project would be inserted into `appmap.yml`. But only packages contained in `module` (and its dependencies) must be inserted.

The old code is taking care of that by filtering the available roots with variable `contentRoots`, which represents all allowed content root directories. AFAIK a project source directory is always located inside a content root.
Because of the possible nesting of modules, the old code retrieved the content roots of the module and all dependency modules and then filtered the available roots with these roots.

IMHO this should work:
```suggestion
        return ProjectRootManager.getInstance(module.getProject()).getModuleSourceRoots(JavaModuleSourceRootTypes.SOURCES)
                .stream()
                .filter(sourceRoot -> VfsUtilCore.isUnder(sourceRoot, contentRoots))
                .toArray(VirtualFile[]::new);
```

Reply:
> With all source roots, all packages available in the project would be inserted into appmap.yml. But only packages contained in module (and its dependencies) must be inserted.

Can you help me understand why this is, please? What's the relationship between IntelliJ's notion of a project and a module, and how is each related to what's actually on disk (e.g. in a git repo)? If you could relate that to what happens when you open https://github.com/finos/waltz in IntelliJ, that would be a big help.

Thanks!

Reply:
In general, usually a single IntelliJ window represents a project. A project consists of 0..n modules and each module has entities, e.g. top-level content roots (directories containing module data), nested source roots (directories containing source files), resource roots, excluded folders, and more. 
Help pages: 
- https://www.jetbrains.com/help/idea/creating-and-managing-modules.html
- https://www.jetbrains.com/help/idea/content-roots.html

The project model is unrelated to a git repo. A project is often imported from an external build system, e.g. Gradle. A Gradle project and its sub-projects is mapped as 1..n modules. It's possible to add several Gradle or Maven projects into a single IntelliJ project.
Therefore we have to be careful about scoping the search for source roots to avoid collecting packages of unrelated modules (e.g. of a different Gradle project or sub-project).

For waltz, you can see all imported modules at `File>Project Structure>Module`. You can select a module to see the content roots, source roots, etc.
- waltz
- waltz-common
- waltz-data
- ...

With this code:
```java
var fileIndex = rootManager.getFileIndex();
        VirtualFile[] contentSourceRoots = rootManager.getContentSourceRoots();
        return Arrays.stream(contentSourceRoots)
                .filter(sourceRoot -> {
                    return fileIndex.isUnderSourceRootOfType(sourceRoot, JavaModuleSourceRootTypes.SOURCES);
                })
                .toArray(VirtualFile[]::new);
```

`rootManager` is a `ProjectRootManager`, which is looking at all modules at once and not just `module`.
A Java package is always located under a source root. If you collect all source roots, then you'll get all of the project's packages.

Reply:
Ok, thanks, I think I understand this explanation.

I'm not sure either of our solutions gets us the result we want, though.

I created this project by importing spring-petclinic as a module in my waltz project:
<img width="1718" alt="image" src="https://github.com/getappmap/appmap-intellij-plugin/assets/4212483/595104e4-d1cd-4eb0-9024-64f026b4e872">

If I run, e.g. the `test` goal in `waltz-jobs`, we want to create `appmap.yml` in the root directory of the local git repo. It should contain all the packages from the all the modules, i.e. the same content as when the `test` goal of the `waltz` module is executed.

Does that make sense?

Reply:
I think we should only use the project-module (i.e. modules, content roots, source roots, etc.) to create appmap.yml and not data like a `.git` directory, which isn't used by the IDE itself to create such model. There may be all kinds of projects without git and the project model contains all we need, I believe.

For waltz, I think I know what's going on.
For Gradle-based multi-module projects (e.g. junit5), the working directory of the Gradle run configuration is the top-level directory. Because it's the top-level dir, the file `appmap.yml` is created in the correct directory.

For waltz, which is Maven-based, the working directory (or, rather, the context module of the run configuration) is the sub-project, e.g. `waltz-jobs`. Therefore `appmap.yml` is created in the sub-directory.
For a checkout of `waltz` at `/sources/waltz`, you can see that:
- top-level module `waltz` has content root `/sources/waltz`
- sub-module `waltz-jobs` has content root `/sources/waltz/waltz-jobs`.

We need changes to the code to make this work for waltz and other setups.

IMHO this should work, using waltz as example
- Find the run configuration context, in our case it's `waltz-jobs` with content root `/sources/waltz/waltz-jobs`
- We would like to create `appmap.yml` in the waltz top-level dir `/sources/waltz`, which belongs to module `waltz`. 
A generic solution would be to iterate the content roots of the project's modules and take the root, which is parent of the start location (e.g. `/sources/waltz/waltz-jobs`) and has no other content root as parent.
In our example, this is `/sources/waltz` of module `waltz`.
- Then the Java packages need to be searched in the module `waltz` plus all modules with a content root below `/sources/waltz` (`waltz-common`, `waltz-data`, `waltz-jobs`, etc.). For correctness, we probably need to look at the dependencies of these modules, too.

I think it would be a small change, but because we had to patch this code a few times already, let's think about it before our next attempt :)

Reply:
Thanks, and I agree about using project information rather than git.

Reply:
Yeah, I only referenced the git repo because I didn't know how else to say that `appmap.yml` needs to get created in the root of the project.